sql/src/test/java/io/crate/metadata/FunctionsTest.java

```diff
@@ -0,0 +1,138 @@
+import io.crate.sql.parser.SqlParser;
+import io.crate.test.integration.CrateUnitTest;
+import io.crate.testing.SQLExecutor;
+import io.crate.types.DataTypes;
+import org.elasticsearch.test.cluster.NoopClusterService;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.Collections;
+
+import static io.crate.metadata.Schemas.*;
+import static io.crate.testing.SymbolMatchers.isLiteral;
+
+public class FunctionsTest extends CrateUnitTest {
+
+    private SQLExecutor e;
+
+    private final FunctionImplementation customFoo = UserDefinedFunctionFactory.of(
+        new UserDefinedFunctionMetaData("custom", "foo",
+            ImmutableList.of(FunctionArgumentDefinition.of(DataTypes.STRING)),
+            DataTypes.STRING,
+            "javascript",
+            "function foo(x) { return x; }")
+    );
+
+    private final FunctionImplementation docLower = UserDefinedFunctionFactory.of(
+        new UserDefinedFunctionMetaData("doc", "lower",
+            ImmutableList.of(FunctionArgumentDefinition.of(DataTypes.STRING)),
+            DataTypes.STRING,
+            "javascript",
+            "function lower(x) { return x; }")
+    );
+
+    @Before
+    public void before() {
+        e = SQLExecutor.builder(new NoopClusterService()).enableDefaultTables().build();
+        e.functions().registerSchemaFunctionResolvers(
+            customFoo.info().ident().schema(),
+            e.functions().generateFunctionResolvers(ImmutableMap.of(customFoo.info().ident(), customFoo))
+        );
+        e.functions().registerSchemaFunctionResolvers(
+            docLower.info().ident().schema(),
+            e.functions().generateFunctionResolvers(ImmutableMap.of(docLower.info().ident(), docLower))
+        );
+    }
+
+    @Test
+    public void testUserDefinedFunctionIgnoresSessionSchema() throws Exception {
+        expectedException.expect(UnsupportedOperationException.class);
+        e.analyzer.boundAnalyze(SqlParser.createStatement("select foo('foo');"),
+            new SessionContext(0, Option.NONE, "custom"),
+            new ParameterContext(Row.EMPTY, Collections.emptyList())).analyzedStatement();
+    }
+
+    @Test
+    public void testBuiltinFunctionIgnoresSessionSchema() throws Exception {
+        SelectAnalyzedStatement analysis = (SelectAnalyzedStatement) e.analyzer.boundAnalyze(
+            SqlParser.createStatement("select abs(1);"),
+            new SessionContext(0, Option.NONE, "bar"),
+            new ParameterContext(Row.EMPTY, Collections.emptyList())).analyzedStatement();
+        assertThat(analysis.relation().querySpec().outputs().get(0), isLiteral(1L));
+    }
+
+    @Test
+    public void testUserDefinedFunctionInCustomSchemaAccessibleViaSchemaName() throws Exception {
+        SelectAnalyzedStatement analysis = (SelectAnalyzedStatement) e.analyzer.boundAnalyze(
+            SqlParser.createStatement("select custom.foo('foo');"),
+            new SessionContext(0, Option.NONE, null),
```

Comment:
please extract all this boilerplate into a method or something to reduce the noise per test